api_callable.go

```diff
@@ -0,0 +1,81 @@
+package gax
+
+import (
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+)
+
+var (
+	retryableGRPCCodes = map[codes.Code]bool{
+		codes.DeadlineExceeded: true,
+		codes.Aborted:          true,
+	}
+)
+
+// Represents a GRPC call stub.
+type APICall func(context.Context, interface{}) (interface{}, error)
+
+// scaleDuration returns the product of a and mult.
+func scaleDuration(a time.Duration, mult float64) time.Duration {
+	ns := float64(a) * mult
+	return time.Duration(int64(ns))
+}
+
+// stubWithRetry returns a wrapper for stub with an exponential backoff retry
+// mechanism based on the values provided in backoffSettings.
+func stubWithRetry(ctx context.Context, stub APICall, backoffSettings backoffSettings) APICall {
+	return func(ctx context.Context, req interface{}) (resp interface{}, err error) {
+		// Forces ctx to expire after a deadline.
+		ctx, _ = context.WithTimeout(ctx, backoffSettings.totalTimeout)
+
+		delay := backoffSettings.delayTimeoutSettings.initial
+		timeout := backoffSettings.rpcTimeoutSettings.initial
+
+		for {
+			// If the deadline is exceeded...
+			if ctx.Err() != nil {
+				return nil, ctx.Err()
+			}
+			timeoutCtx, _ := context.WithTimeout(ctx, backoffSettings.rpcTimeoutSettings.max)
+			timeoutCtx, _ = context.WithTimeout(timeoutCtx, timeout)
+			resp, err = stub(timeoutCtx, req)
+			code := grpc.Code(err)
+			if code == codes.OK {
+				break
+			}
+			if !retryableGRPCCodes[code] {
+				return nil, err
+			}
+			delayCtx, _ := context.WithTimeout(ctx, backoffSettings.delayTimeoutSettings.max)
+			delayCtx, _ = context.WithTimeout(delayCtx, delay)
+			<-delayCtx.Done()
+
+			delay = scaleDuration(delay, backoffSettings.delayTimeoutSettings.multiplier)
+			timeout = scaleDuration(timeout, backoffSettings.rpcTimeoutSettings.multiplier)
+		}
+		return
+	}
+}
+
+// stubWithTimeout returns a wrapper for stub with a timeout applied to its
+// context.
+func stubWithTimeout(ctx context.Context, stub APICall, timeout time.Duration) APICall {
+	return func(ctx context.Context, data interface{}) (interface{}, error) {
+		childCtx, _ := context.WithTimeout(ctx, timeout)
+		return stub(childCtx, data)
+	}
+}
+
+// CreateAPICall returns a wrapper for stub governed by the values provided in
+// settings.
+func CreateAPICall(ctx context.Context, stub APICall, opts ...CallOption) APICall {
+	settings := &callSettings{}
+	callOptions(opts).Resolve(settings)
+	if settings.retrySettings.retryCodes != nil {
```

Comment:
nit: might be good to check the length instead of comparing with nil --- there can be non-nil but empty array, while it's safe to invoke len with nil (see http://play.golang.org/p/3STon08e3T)